File: typescript/src/trials/pepper.ts
/**
 * Per-installation pepper for trial-issuance fingerprint hashing.
 *
 * The hash stored in `trial_issuances.fingerprint_hash` is
 * `SHA-256(pepper || canonical_fingerprint_input)`. Hashing protects raw
 * fingerprints if the table leaks; the pepper makes a stolen table useless
 * outside the deployment that wrote it.
 *
 * The pepper is **operator-managed** — same model as `NUXT_SESSION_PASSWORD`
 * in the admin UI. We deliberately do NOT persist it inside the licensing
 * database, because:
 *
 *   1. A pepper stored alongside the data it protects is just an obfuscation —
 *      anyone who steals the DB has both halves.
 *   2. Operators already have a place for secrets (env vars, secret managers,
 *      KMS); piggybacking on that path matches the rest of the system.
 *
 * The default pepper source is the `LICENSING_TRIAL_PEPPER` environment
 * variable. Tests + dev pass the value in directly.
 *
 * Cross-port byte parity: the Go port computes the same hash with the same
 * pepper, so a pair `(template_id, fingerprint)` deduplicates regardless of
 * which port issues the trial. The interop suite asserts this.
 */

import { createHash } from 'node:crypto';

/** Minimum pepper length (in chars) we accept. 16 bytes of randomness is
 *  the floor for the hash to be unrecoverable in a leaked table. */
const MIN_PEPPER_LENGTH = 32;

/**
 * Read the pepper from `process.env.LICENSING_TRIAL_PEPPER` and validate it.
 * Throws when missing or too short.
 */
export function pepperFromEnv(env = process.env): string {
  const v = env.LICENSING_TRIAL_PEPPER;
  if (typeof v !== 'string' || v.length === 0) {
    throw new Error(
      'LICENSING_TRIAL_PEPPER is required for trial issuance (>= 32 chars; e.g. `openssl rand -hex 32`)',
    );
  }
  if (v.length < MIN_PEPPER_LENGTH) {
    throw new Error(
      `LICENSING_TRIAL_PEPPER must be at least ${MIN_PEPPER_LENGTH} characters (got ${v.length})`,
    );
  }
  return v;
}

/**
 * Hash a canonical fingerprint input with the supplied pepper. The output
 * is lowercase SHA-256 hex, 64 chars — the exact shape persisted in
 * `trial_issuances.fingerprint_hash`.
 */
export function hashFingerprint(pepper: string, fingerprintInput: string): string {
  if (pepper.length < MIN_PEPPER_LENGTH) {
    throw new Error(
      `pepper must be at least ${MIN_PEPPER_LENGTH} characters (got ${pepper.length})`,
    );
  }
  return createHash('sha256').update(pepper).update(fingerprintInput).digest('hex').toLowerCase();
}

/**
 * High-level facade most consumers will use: `Trials.hash(input)` reads the
 * pepper from env once and produces a fingerprint hash. Callers may also
 * construct a `TrialPepperStore` directly to inject a non-env pepper (tests,
 * KMS-resolved secrets).
 */
export class TrialPepperStore {
  readonly #pepper: string;

  constructor(pepper: string) {
    if (pepper.length < MIN_PEPPER_LENGTH) {
      throw new Error(
        `pepper must be at least ${MIN_PEPPER_LENGTH} characters (got ${pepper.length})`,
      );
    }
    this.#pepper = pepper;
  }

  /** Construct from `process.env.LICENSING_TRIAL_PEPPER`. */
  static fromEnv(env: NodeJS.ProcessEnv = process.env): TrialPepperStore {
    return new TrialPepperStore(pepperFromEnv(env));
  }

  /** Hash a canonical fingerprint input. */
  hash(fingerprintInput: string): string {
    return hashFingerprint(this.#pepper, fingerprintInput);
  }
}

File: licensing/trials/pepper.go
// Package trials provides operator-managed pepper handling for the
// trial-issuance fingerprint hash. Mirrors typescript/src/trials/pepper.ts —
// the hash output is byte-identical for a given (pepper, fingerprint) pair
// across both ports, so deduplication works regardless of which port issues
// a trial. Cross-port byte-compat is enforced by the interop test suite.
//
// The pepper is read from the LICENSING_TRIAL_PEPPER environment variable
// by default. We deliberately do not persist it inside the licensing
// database — see the TS doc-comment for the threat model.
package trials

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
)

// MinPepperLength is the minimum pepper length we accept. 32 chars is the
// floor for the digest to be unrecoverable in a leaked table.
const MinPepperLength = 32

// HashFingerprint produces the HMAC-SHA256 hex digest of the fingerprint
// input keyed by the pepper, lowercase, 64 chars — the exact shape
// persisted in trial_issuances.fingerprint_hash.
//
// HMAC rather than the simpler SHA-256(pepper || input): that prefix-MAC
// construction is length-extension-susceptible, so its safety depends on the
// digest never being treated as an authenticator. HMAC makes the primitive
// itself safe, independent of the call site.
func HashFingerprint(pepper, fingerprintInput string) (string, error) {
	if len(pepper) < MinPepperLength {
		return "", fmt.Errorf("pepper must be at least %d characters (got %d)",
			MinPepperLength, len(pepper))
	}
	m := hmac.New(sha256.New, []byte(pepper))
	_, _ = m.Write([]byte(fingerprintInput))
	return hex.EncodeToString(m.Sum(nil)), nil
}

// PepperFromEnv reads the pepper from the LICENSING_TRIAL_PEPPER env var
// (or a caller-supplied lookup function) and validates length.
func PepperFromEnv() (string, error) {
	v := os.Getenv("LICENSING_TRIAL_PEPPER")
	if v == "" {
		return "", fmt.Errorf("LICENSING_TRIAL_PEPPER is required for trial issuance " +
			"(>= 32 chars; e.g. `openssl rand -hex 32`)")
	}
	if len(v) < MinPepperLength {
		return "", fmt.Errorf("LICENSING_TRIAL_PEPPER must be at least %d characters (got %d)",
			MinPepperLength, len(v))
	}
	return v, nil
}

// PepperStore is the high-level facade most consumers will use.
type PepperStore struct {
	pepper string
}

// NewPepperStore validates the pepper length and returns a store ready to hash.
func NewPepperStore(pepper string) (*PepperStore, error) {
	if len(pepper) < MinPepperLength {
		return nil, fmt.Errorf("pepper must be at least %d characters (got %d)",
			MinPepperLength, len(pepper))
	}
	return &PepperStore{pepper: pepper}, nil
}

// PepperStoreFromEnv builds a store from LICENSING_TRIAL_PEPPER.
func PepperStoreFromEnv() (*PepperStore, error) {
	p, err := PepperFromEnv()
	if err != nil {
		return nil, err
	}
	return NewPepperStore(p)
}

// Hash hashes a canonical fingerprint input through the store's pepper.
func (s *PepperStore) Hash(fingerprintInput string) (string, error) {
	return HashFingerprint(s.pepper, fingerprintInput)
}
